diff --git a/toolkit/crashreporter/crash_helper_server/src/logging/env.rs b/toolkit/crashreporter/crash_helper_server/src/logging/env.rs
index 8c9a19109e..189c734a6c 100644
--- a/toolkit/crashreporter/crash_helper_server/src/logging/env.rs
+++ b/toolkit/crashreporter/crash_helper_server/src/logging/env.rs
@@ -33,7 +33,9 @@ fn guess_user_app_data_dir() -> Option<PathBuf> {
 
     cfg_if! {
         if #[cfg(target_os = "linux")] {
-            Some(home_dir.join(".mozilla").join(config::MOZ_APP_NAME))
+            // Use .config/mozilla instead .mozilla
+            let config_dir = dirs::config_dir()?;
+            Some(config_dir.join("mozilla").join(config::MOZ_APP_NAME))
         } else if #[cfg(target_os = "macos")] {
             Some(home_dir.join("Library").join("Application Support").join(config::MOZ_APP_BASENAME))
         } else if #[cfg(target_os = "windows")] {
